File: python/ichika/build_info.py
import datetime
from typing import TypedDict


class RustCInfo(TypedDict):
    """编译器信息"""

    rustc: str
    """RustC 名称"""
    rustc_version: str
    """RustC 版本"""
    opt_level: str
    """优化等级"""
    debug: bool
    """是否为调试编译"""
    jobs: int
    """编译并行数"""


class TargetInfo(TypedDict):
    """编译目标信息"""

    arch: str
    """目标架构"""
    os: str
    """目标操作系统"""
    family: str
    """目标家族"""
    compiler: str
    """使用的编译器"""
    triple: str
    """目标架构标识"""
    endian: str
    """目标端序"""
    pointer_width: str
    """目标指针宽度"""
    profile: str
    """使用的配置"""


class BuildInfo(TypedDict):
    """Ichika 构建信息"""

    builder: RustCInfo
    """Rust 编译器信息"""
    target: TargetInfo
    """编译目标信息"""
    build_time: datetime.datetime
    """构建时间"""
    dependencies: dict
    """构建依赖字典"""
    host_triple: str
    """编译器的架构标识"""

File: src/build_info.rs
use pyo3::prelude::*;
use pyo3::types::PyDict;

use crate::dict;
include!(concat!(env!("OUT_DIR"), "/build-info.rs"));

pub fn get_info(py: Python) -> PyResult<&PyDict> {
    let builder = dict! {py,
        rustc: RUSTC,
        rustc_version: RUSTC_VERSION,
        opt_level: OPT_LEVEL,
        debug: DEBUG,
        jobs: NUM_JOBS,
    };

    let target = dict! {py,
        arch: CFG_TARGET_ARCH,
        os: CFG_OS,
        family: CFG_FAMILY,
        compiler: CFG_ENV,
        triple: TARGET,
        endian: CFG_ENDIAN,
        pointer_width: CFG_POINTER_WIDTH,
        profile: PROFILE,
    };

    let dependencies = PyDict::new(py);
    for (name, ver) in DEPENDENCIES {
        dependencies.set_item(name, ver)?;
    }

    let build_time = py
        .import("email.utils")?
        .getattr("parsedate_to_datetime")?
        .call1((BUILT_TIME_UTC,))?
        .into_py(py);

    Ok(dict! {py,
        builder: builder,
        target: target,
        dependencies: dependencies,
        build_time: build_time,
    })
}
